horovod/_keras/__init__.py

```diff
@@ -77,36 +79,67 @@ def get_gradients(self, loss, params):
             In DistributedOptimizer, get_gradients() is overriden to also
             allreduce the gradients before returning them.
             """
-
-            self._get_gradients_used = True
-            self.grads = super(
-                self.__class__, self).get_gradients(loss, params)
+            self._aggregated_gradients = True
+            if LooseVersion(tf.__version__) >= LooseVersion("2.0.0"):
+                gradients = super(self.__class__, self).get_gradients(loss, params)
+                return self._allreduce(gradients)
+            else:
+                self.grads = super(
+                    self.__class__, self).get_gradients(loss, params)
+                if hvd.size() > 1:
+                    self._agg_helper.init_aggregation_vars(
+                        self.grads,
+                        sess=tf.compat.v1.keras.backend.get_session(op_input_list=()),
+                    )
+                    with tf.control_dependencies([self._profile_helper.profile_start()]):
+                        allreduced_grads = self._agg_helper.compute_gradients(tuple(self.grads))
+                    with tf.control_dependencies(allreduced_grads):
+                        comm_end = self._profile_helper.profile_end()
+                    with tf.control_dependencies([comm_end]):
+                        return [tf.identity(grad) for grad in allreduced_grads]
+                else:
+                    return self.grads
+
+        def _aggregate_gradients(self, grads_and_vars):
+            gradients = [grad for grad, var in grads_and_vars]
+            return self._allreduce(gradients)
+
+        def _allreduce(self, gradients):
+            self._aggregated_gradients = True
             if hvd.size() > 1:
-                self._profile_helper.init_profile_vars(sess=tf.keras.backend.get_session(op_input_list=()))
-                self._agg_helper.init_aggregation_vars(
-                    self.grads,
-                    sess=tf.keras.backend.get_session(op_input_list=()),
-                )
-                with tf.control_dependencies([self._profile_helper.profile_start()]):
-                    allreduced_grads = self._agg_helper.compute_gradients(tuple(self.grads))
-                with tf.control_dependencies(allreduced_grads):
-                    comm_end = self._profile_helper.profile_end()
-                with tf.control_dependencies([comm_end]):
-                    return [tf.identity(grad) for grad in allreduced_grads]
+                averaged_gradients = []
+                with tf.name_scope(self._name + "_Allreduce"):
+                    for grad in gradients:
+                        if grad is not None:
```

Comment:
what do you think about trying to reuse `_make_allreduce_grads_fn` here?

Reply:
I would prefer not to because this is taken directly from the tip of master of Horovod: https://github.com/horovod/horovod/blob/master/horovod/_keras/__init__.py#L52